go-controller/pkg/util/net_linux.go

```diff
@@ -274,74 +278,101 @@ func LinkRoutesDel(link netlink.Link, subnets []*net.IPNet) error {
 	for _, subnet := range subnets {
 		route := &netlink.Route{
 			Dst:       subnet,
 			LinkIndex: link.Attrs().Index,
 			Scope:     netlink.SCOPE_UNIVERSE,
 			Gw:        gwIP,
 		}
+		if len(gwIP) > 0 {
+			route.Gw = gwIP
+		}
+		if len(src) > 0 {
+			route.Src = src
+		}
 		if mtu != 0 {
 			route.MTU = mtu
 		}
 		err := netLinkOps.RouteAdd(route)
 		if err != nil {
-			if os.IsExist(err) {
-				return err
-			}
-			return fmt.Errorf("failed to add route for subnet %s via gateway %s with mtu %d: %v",
-				subnet.String(), gwIP.String(), mtu, err)
+			return fmt.Errorf("failed to add route for subnet %s via gateway %s with mtu %d and src: %s: %v",
+				subnet.String(), gwIP.String(), mtu, src, err)
 		}
 	}
 	return nil
 }
 
-func LinkRoutesAddOrUpdateMTU(link netlink.Link, gwIP net.IP, subnets []*net.IPNet, mtu int) error {
+func LinkRoutesAddOrUpdateSourceOrMTU(link netlink.Link, gwIP net.IP, subnets []*net.IPNet, mtu int, src net.IP) error {
 	for _, subnet := range subnets {
-		route, err := LinkRouteGet(link, gwIP, subnet)
+		route, err := LinkRouteGetFilteredRoute(filterRouteByDst(link, subnet))
 		if err != nil {
 			return err
 		}
 		if route != nil {
-			if route.MTU == mtu {
-				return nil
+			var changed bool
+			if route.MTU != mtu {
+				route.MTU = mtu
+				changed = true
 			}
-			route.MTU = mtu
-			err = netLinkOps.RouteReplace(route)
-			if err != nil {
-				return fmt.Errorf("failed to replace route for subnet %s via gateway %s with mtu %d: %v", subnet.String(), gwIP.String(), mtu, err)
+			if !src.Equal(route.Src) {
+				route.Src = src
+				changed = true
+			}
+
+			if changed {
+				err = netLinkOps.RouteReplace(route)
+				if err != nil {
+					return fmt.Errorf("failed to replace route for subnet %s via gateway %s with mtu %d: %v",
+						subnet.String(), gwIP.String(), mtu, err)
+				}
 			}
 		} else {
-			return LinkRoutesAdd(link, gwIP, []*net.IPNet{subnet}, mtu)
+			return LinkRoutesAdd(link, gwIP, []*net.IPNet{subnet}, mtu, src)
 		}
 	}
 	return nil
 }
 
-// LinkRouteGet gets a route for the given subnet with the specified gwIPStr
+// LinkRouteGetFilteredRoute gets a route for the given route filter.
 // returns nil if route is not found
-func LinkRouteGet(link netlink.Link, gwIP net.IP, subnet *net.IPNet) (*netlink.Route, error) {
-	routeFilter := &netlink.Route{Dst: subnet, LinkIndex: link.Attrs().Index}
-	filterMask := netlink.RT_FILTER_DST | netlink.RT_FILTER_OIF
-	routes, err := netLinkOps.RouteListFiltered(getFamily(gwIP), routeFilter, filterMask)
+func LinkRouteGetFilteredRoute(routeFilter *netlink.Route, filterMask uint64) (*netlink.Route, error) {
+	routes, err := netLinkOps.RouteListFiltered(getFamily(routeFilter.Dst.IP), routeFilter, filterMask)
 	if err != nil {
-		return nil, fmt.Errorf("failed to get routes for subnet %s", subnet.String())
+		return nil, fmt.Errorf(
+			"failed to get routes for subnet %s with filter mask %d: %v",
```

Comment:
Why  not use %v format specifier with the whole filter,  instead of just the filter Dst?

Reply:
so  `... for filter %v and mask %d ...", routeFilter, ...`